### ISSUE TITLE
Automatically add `@updatedAt` attribute where appropriate when `prisma db pull` is run

### ISSUE THREAD

**Issue by jemisonf** [prisma/prisma#15855]
## Problem

If you have a table that looks like:

```sql
CREATE TABLE foo (update_date timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP);
```

The Prisma schema generated by `prisma db pull` will be:

```prisma
model foo {
  update_date DateTime  @default(now()) @db.Timestamp(0)
}
```

The `update_date` column will be set correctly when the record is created but not updated on subsequent queries. This _is_ documented in the Prisma docs, which also identify that you can fix the issue by adding `@updatedAt()` in the Prisma schema. But I don't think this is necessarily obvious enough to avoid missing it if you're porting a large schema into Prisma and have not run into this issue before.

## Suggested solution

If the column in the DB schema has the constraint `ON UPDATE CURRENT_TIMESTAMP` Prisma should automatically add the `@updatedAt` attribute to the corresponding field in the Prisma model when the database schema is pulled.

## Additional context

We're using MySQL 5.6 and Prisma 3.12

**Comment by Karbust** [comment]
I would also like to have the ability to set custom names to the updatedAt column

**Comment by jemisonf** [comment]
> I would also like to have the ability to set custom names to the updatedAt column

@Karbust I might be misunderstanding what you're describing but I think a field with `@updatedAt` should be able to have any name. Might be worth its own issue if there's some reason that doesn't work

**Comment by Karbust** [comment]
> > I would also like to have the ability to set custom names to the updatedAt column
> 
> @Karbust I might be misunderstanding what you're describing but I think a field with `@updatedAt` should be able to have any name. Might be worth its own issue if there's some reason that doesn't work

What I'm talking about is what you are also talking about (if I understood correctly), having to manually add the `@updatedAt` every time you do a database pull. Making prisma automatically assign a certain column as `@updatedAt` would be great. I use `UpdatedAt` on every table, changing every table on each database pull is not the way. Previously I used `sequelize-auto` and I was able to set the name in case I wanted a different name.

I don't think it's needed a new issue, if I understood correctly, your issue is the same as mine.

I'm using PostgreSQL 14.

**Comment by Karbust** [comment]
> I think @jemisonf is asking for Prisma to detect "updated at" characteristics of a column, and add `@updatedAt` based on that automatically.
> 
> If I understand you correctly @Karbust, you want Prisma to react to the name of the column (which would need to be a configuration). By the way, after you add `@updatedAt` once to a column, each further Introspection should _keep_ that attribute on the column. If that is not the case, that would be a bug.

You are absolutely correct, I actually never tried because I assumed it would be overwritten when generating again. Still, would be a good feature to be able to set a config option with the column name

**Comment by janpio** [comment]
Yes - but that would indeed be a different issue than what @jemisonf wants then. Optimally you could open that. Thanks!

**Comment by janpio** [comment]
I think we at Prisma consider the fact that these are using different clocks, as part of the overall problem here. We would definitely not do what this issue asks for before that situation is clarified and potentially fixed (whatever that means) 👍

**Comment by jemisonf** [comment]
> These are not the same thing: ON UPDATE CURRENT_TIMESTAMP is done by the database, even when not going through Prisma, and uses the database server clock. @updatedAt is handled in Prisma, and uses the application server clock. They are not interchangeable and the introspection tool should NOT conflate them.

Thanks @hartbit this makes perfect sense!

To step back a little bit with this problem: the core issue for me is that as someone migrating to Prisma my expectation was that the contract I had in my old code (if I don't pass in a value for `updated_at` when updating the record it's automatically set to the current date according to the database clock) will continue to kept in my Prisma code without extra work on my part. You can more or less do this by adding `@updatedAt` to the fields (for our purposes we don't really care about the difference between app and db clocks) but it's not immediately obvious that that's necessary, particularly if you're importing a large schema, and in our case we did it only after noticing `update_date` wasn't getting updated correctly for some of our records.

@janpio I'm not sure I have any good ideas for how to solve that -- the best I could come up with is creating a new decoration like `@dbUpdatedAt` in addition to `@updatedAt` -- but I do think it might be worth trying to come up with a solution to this issue.

**Comment by hartbit** [comment]
> @janpio I'm not sure I have any good ideas for how to solve that -- the best I could come up with is creating a new decoration like @dbUpdatedAt in addition to @updatedAt -- but I do think it might be worth trying to come up with a solution to this issue.

The lack of this feature is our number 1 issue with Prisma at the moment. For information, there is already an issue tracking this feature request: https://github.com/prisma/prisma/issues/9317. Until we have that, if we want to have the database update the field, there is no choice but to set `ON UPDATE CURRENT_TIMESTAMP` manually in the migrations because of the following bug: https://github.com/prisma/prisma/issues/11374

**Comment by janpio** [comment]
@hartbit Did you see https://github.com/prisma/prisma/issues/11374? It implies that one might be able to set the value via `dbgenerated()` (as long as one does not rely on `db pull` to update the schema later on, or will manually make sure to keep the values).

**Comment by hartbit** [comment]
@janpio Yes, I saw that, but we rely on `db pull` to prototype changes in between migrations. It's cumbersome to remind developers to check that they do not commit the removal of the `dbgenerated` attribute.

**Comment by wmtrinu** [comment]
Would "server clock/timezone" matter if using the type of `timestamp with time zone`? (postgres)

**Comment by mqliutie** [comment]
So when row data is changed, Developers have to go through the changes manually?
The updateAt (or whatever it's called) auto-update feature comes in handy if a row or column changes!

**Comment by gxanshu** [comment]
also when we push with `@updateAt` it should add `ON UPDATE CURRENT_TIMESTAMP` in db.

**Comment by ludralph** [comment]
Closing this in favour of [12076](https://github.com/prisma/prisma/issues/12076)
